libwasmvm/src/test_utils.rs

```diff
@@ -0,0 +1,89 @@
+#![cfg(test)]
+
+use cosmwasm_std::{Decimal, Uint128};
+use std::str::FromStr as _;
+
+/// Asserts that two expressions are approximately equal to each other.
+///
+/// The ratio argument defines how wide a range of values we accept, and is applied
+/// to the **second** (`right`) argument.
+///
+/// On panic, this macro will print the values of the expressions with their
+/// debug representations, and info about the acceptable range.
+///
+/// Like [`assert_eq!`], this macro has a second form, where a custom
+/// panic message can be provided.
+#[macro_export]
+macro_rules! assert_approx_eq {
+    ($left:expr, $right:expr, $ratio:expr $(,)?) => {{
+        $crate::test_utils::assert_approx_eq_impl($left, $right, $ratio, None);
+    }};
+    ($left:expr, $right:expr, $ratio:expr, $($args:tt)+) => {{
+        $crate::test_utils::assert_approx_eq_impl($left, $right, $ratio, Some(format!($($args)*)));
+    }};
+}
+
+#[track_caller]
+fn assert_approx_eq_impl(
+    left: impl Into<Uint128>,
+    right: impl Into<Uint128>,
+    ratio: &str,
+    panic_msg: Option<String>,
+) {
+    let left = left.into();
+    let right = right.into();
+    let ratio = Decimal::from_str(ratio).unwrap();
+
+    let delta = right * ratio;
+    let lower_bound = right - delta;
+    let upper_bound = right + delta;
+
+    if !(left >= lower_bound && left <= upper_bound) {
```

Comment:
You could also calculate `let actual_delta = left.abs_diff(right)`. Not sure if it helps or makes things nicer.

Reply:
I haven't thought of that. Sounds better.